op-chain-ops/script/check_super_root.go

```
@@ -0,0 +1,379 @@
+
+	for url, client := range m.ethClients {
+		m.log.Debug("Fetching finalized block header", "url", url)
+		header, err := client.HeaderByNumber(ctx, nil)
+		if err != nil {
+			return fmt.Errorf("failed to get finalized header from %s: %w", url, err)
+		}
+		if header == nil {
+			return fmt.Errorf("received nil finalized header from %s", url)
+		}
+		m.log.Debug("Got finalized header", "url", url, "number", header.Number, "timestamp", header.Time)
+
+		if minTimestamp == nil || header.Time < *minTimestamp {
+			timestamp := header.Time
+			minTimestamp = &timestamp
+			m.log.Debug("Updated minimum timestamp", "url", url, "new_min_timestamp", *minTimestamp)
+		}
+	}
+
+	if minTimestamp == nil {
+		return errors.New("no valid finalized timestamps found across connected chains")
+	}
+	m.anchorTimestamp = *minTimestamp
+	m.log.Info("Determined common anchor timestamp", "timestamp", m.anchorTimestamp)
+	return nil
+}
+
+// deriveParamsAndFindTargetBlocks derives block time, estimates genesis timestamp,
+// and finds the specific block at or just before the anchor timestamp for each chain.
+func (m *SuperRootMigrator) deriveParamsAndFindTargetBlocks(ctx context.Context) error {
+	m.log.Info("Deriving chain parameters and finding target blocks...")
+
+	for url, settings := range m.chainSettings {
+		client := m.ethClients[url]
+		m.log.Debug("Processing chain", "url", url, "chain_id", settings.ChainID)
+
+		finalizedHeader, err := client.HeaderByNumber(ctx, nil)
+		if err != nil {
+			return fmt.Errorf("failed to get finalized block header for %s: %w", url, err)
+		}
+		if finalizedHeader == nil || finalizedHeader.Number == nil {
+			return fmt.Errorf("received invalid finalized header for %s", url)
+		}
+		if finalizedHeader.Number.Sign() <= 0 {
+			// Cannot derive block time if finalized is genesis or block 0
+			m.log.Warn("Finalized block is genesis or block 0, cannot derive block time. Using fallback 2s.", "url", url)
+			settings.BlockTime = 2
```

It's not safe to fallback to 2s block time - we should just return an error if we can't calculate the actual block time.